README.md
# RemoteRadioThermostat

Android Things app that uses OneWire Ds18b20 temperature sensor and
RadioThermostat CT-50 Wi-Fi Thermostat to measure temperature and
update Firebase realtime database for remote tracking.


onewire/CHANGELOG.md
# Change Log

## [0.1] - 2018-01-14
- initial version


app/src/test/java/org/efimov/remoteradiothermostat/ExampleUnitTest.java
package org.efimov.remoteradiothermostat;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

onewire/src/test/java/com/google/android/things/contrib/driver/onewire/Ds18b20Test.java
/*
 * Copyright 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.onewire;

import com.google.android.things.pio.UartDevice;

import junit.framework.Assert;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.mockito.verification.VerificationMode;

import java.io.IOException;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.byteThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.times;

public class Ds18b20Test {

    @Mock
    private UartDevice mUart;

    @Rule
    public MockitoRule mMockitoRule = MockitoJUnit.rule();

    @Rule
    public ExpectedException mExpectedException = ExpectedException.none();

    @Test
    public void close() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        ds18b20.close();
        Mockito.verify(mUart).close();
    }

    @Test
    public void close_safeToCallTwice() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        ds18b20.close();
        ds18b20.close();  // should not throw
        Mockito.verify(mUart).close();
    }

    @Test
    public void readTemperature_throwsIfClosed() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        ds18b20.close();
        mExpectedException.expect(IllegalStateException.class);
        mExpectedException.expectMessage("not open");
        ds18b20.readTemperature();
    }

    @Test
    public void readTemperature_throwsIfNoDevices() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        // Initial Setup
        Mockito.verify(mUart).setDataSize(8);
        Mockito.verify(mUart).setParity(0);
        Mockito.verify(mUart).setStopBits(1);
        Mockito.verify(mUart).setHardwareFlowControl(0);

        Mockito.when(mUart.read(any(byte[].class), eq(1))).thenReturn(1);
        mExpectedException.expect(IOException.class);
        mExpectedException.expectMessage("not found");
        ds18b20.readTemperature();
        Mockito.verify(mUart).setBaudrate(9600);
        Mockito.verify(mUart).write(new byte[]{(byte) (0xf0)}, 1);
        Mockito.verify(mUart).setBaudrate(115200);
    }

    @Test
    public void oneWire_ResetFails() throws IOException {
        OneWire oneWire = new OneWire(mUart);
        // Initial Setup
        Mockito.verify(mUart).setDataSize(8);
        Mockito.verify(mUart).setParity(0);
        Mockito.verify(mUart).setStopBits(1);
        Mockito.verify(mUart).setHardwareFlowControl(0);
        // Reset fails because read is returning '0'.
        Mockito.when(mUart.read(any(byte[].class), eq(1))).thenReturn(1);
        mExpectedException.expect(IOException.class);
        mExpectedException.expectMessage("not found");
        boolean success = oneWire.reset();
        Assert.assertFalse(success);
        Mockito.verify(mUart).setBaudrate(9600);
        Mockito.verify(mUart).write(new byte[]{(byte) (0xf0)}, 1);
        Mockito.verify(mUart).setBaudrate(115200);
    }

    @Test
    public void convertTemperatureValid() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        // Valid raw measurement
        byte[] validRaw = {0x3e, 0x01, 0x4b, 0x46, 0x7f, (byte) 0xff, 0x0c, 0x10, (byte) 0xb0};
        float temp = ds18b20.convertTemperature(validRaw);
        assertEquals(19.875f, temp);
        assertEquals(85f,
                ds18b20.convertTemperature(new byte[]{(byte) 0x50, (byte) 0x05, 0, 0, 0, 0, 0, 0, 0}));
        assertEquals(125f,
                ds18b20.convertTemperature(new byte[]{(byte) 0xd0, (byte) 0x07, 0, 0, 0, 0, 0, 0, 0}));
    }

    @Test
    public void convertTemperatureZero() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        // Raw measurement equal to 0 celsius (but with invalid CRC8)
        byte[] validRaw = {0x0, 0x0, 0, 0, 0, 0, 0, 0, 0};
        float temp = ds18b20.convertTemperature(validRaw);
        assertEquals(0f, temp);
    }

    @Test
    public void convertTemperatureNegative() throws IOException {
        Ds18b20 ds18b20 = new Ds18b20(mUart);
        // Raw measurement equal to -1/16 celsius (but with invalid CRC8)
        byte[] validRaw = {(byte) 0x5e, (byte) 0xff, 0, 0, 0, 0, 0, 0, 0};
        float temp = ds18b20.convertTemperature(validRaw);
        assertEquals(-10.125f, temp);
        assertEquals(-25.0625f,
                ds18b20.convertTemperature(new byte[]{(byte) 0x6f, (byte) 0xfe, 0, 0, 0, 0, 0, 0, 0}));
        assertEquals(-0.5f,
                ds18b20.convertTemperature(new byte[]{(byte) 0xf8, (byte) 0xff, 0, 0, 0, 0, 0, 0, 0}));

    }
}


radiothermostat/src/test/java/com/google/android/things/contrib/driver/radiothermostat/RadioThermostatTest.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.radiothermostat;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.HttpURLConnection;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.times;

public class RadioThermostatTest {

    @Mock
    private HttpURLConnection mConnection;

    @Mock
    private JSONObject mResponseJson;

    @Rule
    public MockitoRule mMockitoRule = MockitoJUnit.rule();

    @Rule
    public ExpectedException mExpectedException = ExpectedException.none();

    @Test
    public void close() throws IOException {
        RadioThermostat radioThermostat = new RadioThermostat(mConnection);
        radioThermostat.close();
        Mockito.verify(mConnection).disconnect();
    }

    @Test
    public void close_safeToCallTwice() throws IOException {
        RadioThermostat radioThermostat = new RadioThermostat(mConnection);
        radioThermostat.close();
        radioThermostat.close(); // should not throw
        Mockito.verify(mConnection, times(1)).disconnect();
    }

    @Test
    public void setMode() throws IOException {
        RadioThermostat radioThermostat = new RadioThermostat(mConnection);
        radioThermostat.setMode(RadioThermostat.MODE_NORMAL);
        Mockito.reset(mConnection);
        radioThermostat.setMode(RadioThermostat.MODE_SLEEP);
    }

    @Test
    public void setMode_throwsIfClosed() throws IOException {
        RadioThermostat radioThermostat = new RadioThermostat(mConnection);
        radioThermostat.close();
        mExpectedException.expect(IllegalStateException.class);
        radioThermostat.setMode(RadioThermostat.MODE_NORMAL);
    }

    @Test
    public void readTemperature() throws IOException, JSONException {
        String response = "{\"temp\":66.00,\"tmode\":1,\"fmode\":2,\"override\":1,\"hold\":0,\"t_heat\":45.00,\"tstate\":0,\"fstate\":1,\"time\":{\"day\":2,\"hour\":19,\"minute\":57},\"t_type_post\":0}";
        ByteArrayInputStream data = new ByteArrayInputStream(response.getBytes());
        Mockito.when(mConnection.getInputStream()).thenReturn(data);
        Mockito.when(mResponseJson.getDouble("temp")).thenReturn((double)66f);
        RadioThermostat radioThermostat = new RadioThermostat(mConnection, mResponseJson);

        float temperature = radioThermostat.readTemperature();
        assertEquals(18.88889f, temperature);
    }

    @Test
    public void readTemperature_throwsIfClosed() throws IOException {
        RadioThermostat radioThermostat = new RadioThermostat(mConnection);
        radioThermostat.close();
        mExpectedException.expect(IllegalStateException.class);
        mExpectedException.expectMessage("not open");
        radioThermostat.readTemperature();
    }

}


app/src/main/java/org/efimov/remoteradiothermostat/MainActivity.java
package org.efimov.remoteradiothermostat;

import android.app.Activity;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventCallback;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.os.StrictMode;
import android.util.Log;

import com.google.android.things.contrib.driver.onewire.Ds18b20SensorDriver;
import com.google.android.things.contrib.driver.radiothermostat.RadioThermostatSensorDriver;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ServerValue;

import java.io.IOException;


public class MainActivity extends Activity {
    private static final String TAG = MainActivity.class.getSimpleName();

    private SensorManager mSensorManager;

    private RadioThermostatSensorDriver mRadioThermostatSensorDriver;
    private Ds18b20SensorDriver mDs18b20SensorDriver;

    private FirebaseDatabase mDatabase;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Connect to Firebase database.
        mDatabase = FirebaseDatabase.getInstance();

        mSensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
        mSensorManager.registerDynamicSensorCallback(new SensorCallback());

        try {
            // TODO(mef): FIX THIS
            //StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder().permitAll().build());
            //mRadioThermostatSensorDriver = new RadioThermostatSensorDriver("192.168.1.60");
            //mRadioThermostatSensorDriver.registerTemperatureSensor();
            mDs18b20SensorDriver = new Ds18b20SensorDriver("UART6");
            mDs18b20SensorDriver.registerTemperatureSensor();
        } catch (IOException e) {
            Log.e(TAG, "Cannot register sensor", e);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    private class SensorListener extends SensorEventCallback {
        @Override
        public void onSensorChanged (SensorEvent event) {
            Log.e("SensorEventCallback", "onSensorChanged: " + event.sensor.toString());
            Log.e("SensorEventCallback", "Sensor Value: " + event.values[0]);
            String sensorName = event.sensor.getName();
            final DatabaseReference log = mDatabase.getReference("Temperatures").child(sensorName);
            // upload temperature to firebase.
            log.child("timestamp").setValue(ServerValue.TIMESTAMP);
            log.child("sensor").setValue(sensorName);
            log.child("temp").setValue(event.values[0]);
        }
    }

    private class SensorCallback extends SensorManager.DynamicSensorCallback {
        @Override
        public void onDynamicSensorConnected(Sensor sensor) {
            //Sensor connected
            // mSensorManager?.registerListener(this@SensorDriverService, sensor,
            //        SensorManager.SENSOR_DELAY_NORMAL)
            Log.e("DynamicSensorCallback", "Sensor Connected. " + sensor.toString());
            if (sensor.getType() == Sensor.TYPE_AMBIENT_TEMPERATURE) {
                mSensorManager.registerListener(new SensorListener(), sensor, SensorManager.SENSOR_DELAY_NORMAL);
            }
        }

        @Override
        public void onDynamicSensorDisconnected(Sensor sensor) {
            //Sensor disconnected
            // mSensorManager.unregisterListener(this);
            Log.e("DynamicSensorCallback", "Sensor Disconnected. " + sensor.toString());
        }
    }

}


onewire/src/main/java/com/google/android/things/contrib/driver/onewire/Ds18b20.java
/*
 * Copyright 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.onewire;

import android.support.annotation.VisibleForTesting;

import com.google.android.things.pio.UartDevice;

import java.io.IOException;

/**
 * Driver for the DS18B20 temperature sensor.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class Ds18b20 extends OneWire {

    private static final String TAG = Ds18b20.class.getSimpleName();

    static final int DS18X20_CONVERT_T = 0x44;
    static final int DS18X20_READ = 0xBE;

    // Sensor constants from the datasheet.
    // https://datasheets.maximintegrated.com/en/ds/DS18B20.pdf

    /**
     * Mininum temperature in Celsius the sensor can measure.
     */
    public static final float MIN_TEMP_C = -55f;
    /**
     * Maximum temperature in Celsius the sensor can measure.
     */
    public static final float MAX_TEMP_C = 125f;
    /**
     * Maximum power consumption in micro-amperes when measuring temperature.
     */
    public static final float MAX_POWER_CONSUMPTION_TEMP_UA = 1500f;

    /**
     * Maximum frequency of the measurements.
     */
    public static final float MAX_FREQ_HZ = 1f;
    /**
     * Minimum frequency of the measurements.
     */
    public static final float MIN_FREQ_HZ = 1/600f;

    /**
     * Create a new Ds18b20 sensor driver connected on the given UART.
     *
     * @param uart UART port the sensor is connected to.
     * @throws IOException
     */
    public Ds18b20(String uart) throws IOException {
        super(uart);
    }

    /**
     * Create a new Ds18b20 sensor driver connected on the given UART with particular ID.
     *
     * @param uart UART port the sensor is connected to.
     * @param id   OneWire ID of the sensor.
     * @throws IOException
     */
    public Ds18b20(String uart, String id) throws IOException {
        super(uart, id);
    }

    /**
     * Create a new OneWire sensor driver connected on the given UART.
     *
     * @param device UART device of the sensor.
     * @throws IOException
     */
    @VisibleForTesting
    /*package*/ Ds18b20(UartDevice device) throws IOException {
        super(device);
    }

    /**
     * Create a new OneWire sensor driver connected on the given UART with particular ID..
     *
     * @param device UART device of the sensor.
     * @param id     OneWire ID of the sensor.
     * @throws IOException
     */
    @VisibleForTesting
    /*package*/ Ds18b20(UartDevice device, String id) throws IOException {
        super(device, id);
    }

    /**
     * Read the current temperature.
     *
     * @return the current temperature in degrees Celsius
     */
    float readTemperature() throws IOException {
        oneWireCommand(DS18X20_CONVERT_T, getOneWireId());
        // Wait for conversion
        try {
            while (oneWireBit(true)) {
                Thread.sleep(100);
            }
        } catch (InterruptedException e) {
            throw new IOException("Interrupted waiting for conversion.");
        }
        // Read result
        oneWireCommand(DS18X20_READ, getOneWireId());
        return convertTemperature(oneWireReadBytes(9));
    }

    float convertTemperature(byte[] rawMeasure) {
        // TODO(mef): Verify CRC8 of the result.
        int msb = rawMeasure[1];
        int lsb = rawMeasure[0];
        if (lsb < 0) {
            lsb = 256 + lsb;
        }
        float tempRead = msb * 256 + lsb;
        float temp = tempRead / 16;
        return temp;
    }
}


onewire/src/main/java/com/google/android/things/contrib/driver/onewire/Ds18b20SensorDriver.java
/*
 * Copyright 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.onewire;

import android.hardware.Sensor;

import com.google.android.things.userdriver.UserDriverManager;
import com.google.android.things.userdriver.UserSensor;
import com.google.android.things.userdriver.UserSensorDriver;
import com.google.android.things.userdriver.UserSensorReading;

import java.io.IOException;
import java.util.UUID;

public class Ds18b20SensorDriver implements AutoCloseable {
    private static final String TAG = "Ds18b20SensorDriver";

    // DRIVER parameters
    // documented at https://source.android.com/devices/sensors/hal-interface.html#sensor_t
    private static final String DRIVER_VENDOR = "Dallas Semiconductors";
    private static final String DRIVER_NAME = "DS18B20";
    private static final int DRIVER_MIN_DELAY_US = Math.round(1000000.f / Ds18b20.MAX_FREQ_HZ);
    private static final int DRIVER_MAX_DELAY_US = Math.round(1000000.f / Ds18b20.MIN_FREQ_HZ);

    private Ds18b20 mDevice;

    private TemperatureUserDriver mTemperatureUserDriver;

    /**
     * Create a new Ds18b20 sensor driver connected on the given UART.
     * The driver emits {@link android.hardware.Sensor} with temperature data when
     * registered.
     * @param uart UART port the sensor is connected to.
     * @throws IOException
     * @see #registerTemperatureSensor()
     */
    public Ds18b20SensorDriver(String uart) throws IOException {
        mDevice = new Ds18b20(uart);
    }

    /**
     * Close the driver and the underlying device.
     * @throws IOException
     */
    @Override
    public void close() throws IOException {
        unregisterTemperatureSensor();
        if (mDevice != null) {
            try {
                mDevice.close();
            } finally {
                mDevice = null;
            }
        }
    }

    /**
     * Register a {@link UserSensor} that pipes temperature readings into the Android SensorManager.
     * @see #unregisterTemperatureSensor()
     */
    public void registerTemperatureSensor() {
        if (mDevice == null) {
            throw new IllegalStateException("cannot register closed driver");
        }

        if (mTemperatureUserDriver == null) {
            mTemperatureUserDriver = new TemperatureUserDriver();
            UserDriverManager.getManager().registerSensor(mTemperatureUserDriver.getUserSensor());
        }
    }

    /**
     * Unregister the temperature {@link UserSensor}.
     */
    public void unregisterTemperatureSensor() {
        if (mTemperatureUserDriver != null) {
            UserDriverManager.getManager().unregisterSensor(mTemperatureUserDriver.getUserSensor());
            mTemperatureUserDriver = null;
        }
    }

    private class TemperatureUserDriver extends UserSensorDriver {
        // DRIVER parameters
        // documented at https://source.android.com/devices/sensors/hal-interface.html#sensor_t
        private static final float DRIVER_MAX_RANGE = Ds18b20.MAX_TEMP_C;
        private static final float DRIVER_RESOLUTION = 0.5f;
        private static final float DRIVER_POWER = Ds18b20.MAX_POWER_CONSUMPTION_TEMP_UA / 1000.f;
        private static final int DRIVER_VERSION = 1;
        private static final String DRIVER_REQUIRED_PERMISSION = "";

        private boolean mEnabled;
        private UserSensor mUserSensor;

        private UserSensor getUserSensor() {
            if (mUserSensor == null) {
                mUserSensor = new UserSensor.Builder()
                        .setType(Sensor.TYPE_AMBIENT_TEMPERATURE)
                        .setName(DRIVER_NAME)
                        .setVendor(DRIVER_VENDOR)
                        .setVersion(DRIVER_VERSION)
                        .setMaxRange(DRIVER_MAX_RANGE)
                        .setResolution(DRIVER_RESOLUTION)
                        .setPower(DRIVER_POWER)
                        .setMinDelay(DRIVER_MIN_DELAY_US)
                        .setRequiredPermission(DRIVER_REQUIRED_PERMISSION)
                        .setMaxDelay(DRIVER_MAX_DELAY_US)
                        .setUuid(UUID.randomUUID())
                        .setDriver(this)
                        .build();
            }
            return mUserSensor;
        }

        @Override
        public UserSensorReading read() throws IOException {
            return new UserSensorReading(new float[]{mDevice.readTemperature()});
        }

        @Override
        public void setEnabled(boolean enabled) throws IOException {
            mEnabled = enabled;
        }

        private boolean isEnabled() {
            return mEnabled;
        }
    }

}


radiothermostat/src/main/java/com/google/android/things/contrib/driver/radiothermostat/RadioThermostat.java
/*
 * Copyright 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.radiothermostat;

import android.support.annotation.IntDef;
import android.support.annotation.VisibleForTesting;

import com.google.android.things.pio.I2cDevice;
import com.google.android.things.pio.PeripheralManagerService;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 * Driver for the RadioThermostat temperature sensor.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class RadioThermostat implements AutoCloseable {

    private static final String TAG = RadioThermostat.class.getSimpleName();

    // Sensor constants from the datasheet.
    // https://radiothermostat.desk.com/customer/portal/articles/1268461-where-do-i-find-information-about-the-wifi-api-
    /**
     * Mininum temperature in Celsius the sensor can measure.
     */
    public static final float MIN_TEMP_C = -40f;
    /**
     * Maximum temperature in Celsius the sensor can measure.
     */
    public static final float MAX_TEMP_C = 85f;
    /**
     * Temperature reported if sensor is invalid.
     */
    public static final float INVALID_TEMP_C = -99f;
    /**
     * Maximum power consumption in micro-amperes when measuring temperature.
     */
    public static final float MAX_POWER_CONSUMPTION_TEMP_UA = 325f;
    /**
     * Maximum frequency of the measurements.
     */
    public static final float MAX_FREQ_HZ = 1f;
    /**
     * Minimum frequency of the measurements.
     */
    public static final float MIN_FREQ_HZ = 0.01f;

    /**
     * Power mode.
     */
    @Retention(RetentionPolicy.SOURCE)
    @IntDef({MODE_SLEEP, MODE_FORCED, MODE_NORMAL})
    public @interface Mode {
    }

    public static final int MODE_SLEEP = 0;
    public static final int MODE_FORCED = 1;
    public static final int MODE_NORMAL = 2;

    private HttpURLConnection mConnnection;
    private JSONObject mLastResponse;
    private int mMode = MODE_NORMAL;
    private boolean mEnabled;
    private float mLastTemperature = INVALID_TEMP_C;

    /**
     * Create a new Radio Thermostat sensor driver connected on the given host.
     *
     * @param host Name or IP Address of radio thermostat host.
     */
    public RadioThermostat(String host) throws IOException {
        URL urlObj = new URL("http://" + host + "/tstat");
        mConnnection = (HttpURLConnection) urlObj.openConnection();
    }

    @VisibleForTesting
    /*package*/ RadioThermostat(HttpURLConnection connection) {
        mConnnection = connection;
    }

    @VisibleForTesting
    /*package*/ RadioThermostat(HttpURLConnection connection, JSONObject responseJson) {
        mConnnection = connection;
        mLastResponse = responseJson;
    }

    private void connect() throws IOException {
        mConnnection.setDoOutput(false);
        mConnnection.setRequestMethod("GET");
        mConnnection.setConnectTimeout(15000);
        mConnnection.connect();
    }

    private void getTemperatureFromThermostat() throws IOException, JSONException {
        connect();
        //Receive the response from the server
        InputStream in = new BufferedInputStream(mConnnection.getInputStream());
        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
        StringBuilder result = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            result.append(line);
        }
        in.close();
        reader.close();
        if (mLastResponse == null) {
            mLastResponse = new JSONObject(result.toString());
        }
    }

    /**
     * Read the current temperature.
     *
     * @return the current temperature in degrees Celsius
     */
    private float readTemperatureJson() throws IOException, IllegalStateException {
        try {
            // try parse the string to a JSON object
            getTemperatureFromThermostat();
            double tempF = mLastResponse.getDouble("temp");
            return (float) (tempF - 32.0f) / 1.8f;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return INVALID_TEMP_C;
    }

    /**
     * Close the driver and the underlying connection.
     */
    @Override
    public void close() throws IOException {
        if (mConnnection != null) {
            mConnnection.disconnect();
            mConnnection = null;
        }
    }

    public void setMode(int mode) throws IllegalStateException {
        if (mConnnection == null) {
            throw new IllegalStateException("Thermostat connection is not open.");
        }
        mMode = mode;
    }

    /**
     * Read the current temperature.
     *
     * @return the current temperature in degrees Celsius
     */
    public float readTemperature() throws IOException, IllegalStateException {
        if (mConnnection == null) {
            throw new IllegalStateException("Thermostat connection is not open.");
        }
        switch (mMode) {
            case MODE_FORCED:
                return readTemperatureJson();
            case MODE_NORMAL:
                return readTemperatureJson();
            case MODE_SLEEP:
                return mLastTemperature;
        }
        return mLastTemperature;
    }
}


onewire/src/main/java/com/google/android/things/contrib/driver/onewire/OneWire.java
/*
 * Copyright 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.things.contrib.driver.onewire;

import android.support.annotation.VisibleForTesting;
import android.util.Log;

import com.google.android.things.pio.PeripheralManagerService;
import com.google.android.things.pio.UartDevice;

import java.io.IOException;

public class OneWire implements AutoCloseable {
    private static final String TAG = OneWire.class.getSimpleName();

    static final byte OW_MATCH_ROM = 0x55;
    static final byte OW_SKIP_ROM = (byte) 0xcc;
    static final byte OW_SEARCH_ROM = (byte) 0xf0;

    UartDevice mUartDevice;
    byte[] mOneWireId;

    /**
     * Create a new OneWire sensor driver connected on the given UART.
     *
     * @param uart UART port the sensor is connected to.
     * @throws IOException
     */
    public OneWire(String uart) throws IOException {
        this(uart, null);
    }

    /**
     * Create a new OneWire sensor driver connected on the given UART with particular ID.
     *
     * @param uart UART port the sensor is connected to.
     * @param id   OneWire ID of the sensorr.
     * @throws IOException
     */
    public OneWire(String uart, String id) throws IOException {
        this(new PeripheralManagerService().openUartDevice(uart), id);
    }

    /**
     * Create a new OneWire sensor driver connected on the given UART.
     *
     * @param device UART device of the sensor.
     * @throws IOException
     */
    @VisibleForTesting
    /*package*/ OneWire(UartDevice device) throws IOException {
        this(device, null);
    }

    /**
     * Create a new OneWire sensor driver connected on the given UART with particular ID..
     *
     * @param device UART device of the sensor.
     * @param id     OneWire ID of the sensor.
     * @throws IOException
     */
    @VisibleForTesting
    /*package*/ OneWire(UartDevice device, String id) throws IOException {
        mUartDevice = device;

        try {
            mUartDevice.setDataSize(8);
            mUartDevice.setParity(UartDevice.PARITY_NONE);
            mUartDevice.setStopBits(1);
            mUartDevice.setHardwareFlowControl(UartDevice.HW_FLOW_CONTROL_NONE);
            // This hangs 'close' test if there is no data to read.
            // reset();
        } catch (IOException | RuntimeException e) {
            try {
                close();
            } catch (IOException | RuntimeException ignored) {
            }
            throw e;
        }
    }

    /**
     * Returns the One Wire Device ID.
     */
    public byte[] getOneWireId() {
        return mOneWireId;
    }

    protected boolean oneWireBit(boolean b) throws IOException {
        if (b) {
            /* Write 1 */
            uartWriteByte(0xff);
        } else {
            /* Write 0 */
            uartWriteByte(0x00);
        }

        /* Read */
        int c = uartReadByte();
        b = ((c & 0xff) == 0xff);
        return b;
    }

    protected byte oneWireWriteByte(byte b) throws IOException {
        int i = 8;
        do {
            boolean j = oneWireBit((b & 1) != 0);
            b >>= 1;
            b &= 0x7f;
            if (j) {
                b |= 0x80;
            }
        } while (--i != 0);
        return b;
    }

    // Read bytes by writing 0xFF.
    protected byte[] oneWireReadBytes(int readCount) throws IOException {
        byte[] bytes = new byte[readCount];
        for (int i = 0; i < readCount; ++i) {
            bytes[i] = oneWireWriteByte((byte) 0xff);
        }
        return bytes;
    }

    void oneWireCommand(int command, byte[] id) throws IOException {
        reset();
        if (id != null) {
            oneWireWriteByte(OW_MATCH_ROM);
            for (byte b : id) {
                oneWireWriteByte(b);
            }
        } else {
            oneWireWriteByte(OW_SKIP_ROM);
        }
        oneWireWriteByte((byte) command);
    }

    private void uartWriteByte(int b) throws IOException {
        if (mUartDevice == null) {
            throw new IllegalStateException("Uart device is not open");
        }
        byte[] buffer = new byte[]{(byte) b};
        mUartDevice.write(buffer, 1);
    }

    private int uartReadByte() throws IOException {
        if (mUartDevice == null) {
            throw new IllegalStateException("Uart device is not open");
        }
        // Maximum amount of data to read at one time
        final int maxCount = 1;
        byte[] buffer = new byte[maxCount];
        while (mUartDevice.read(buffer, buffer.length) == 0) {
            try {
                Thread.sleep(5);
            } catch (InterruptedException e) {
                break;
            }
        }
        int b = (buffer[0] & 0xff);
        return b;
    }

    protected boolean reset() throws IOException {
        if (mUartDevice == null) {
            throw new IllegalStateException("Uart device is not open");
        }
        mUartDevice.setBaudrate(9600);
        uartWriteByte(0xf0);
        int probe = uartReadByte();
        mUartDevice.setBaudrate(115200);
        if (probe == 0 || probe == 0xf0) {
            throw new IOException("OneWire devices not found");
        }
        return true;
    }

    @Override
    public void close() throws IOException {
        if (mUartDevice != null) {
            try {
                mUartDevice.close();
                mUartDevice = null;
            } catch (IOException e) {
                Log.w(TAG, "Unable to close UART device", e);
            }
        }
    }
}
